src/pgn.js

```diff
@@ -315,31 +327,379 @@ function gameGetterImpl(impl, gameIndex) {
-			if(!doSkipGame(stream)) {
-				break;
+			try {
+				if(!checkValidGame(stream)) {
+					break;
+				}
+				games.push(currentPos);
+			}
+			catch (err) {
+				if (err instanceof exception.InvalidPGN) {
+					errors.push({message: err.message, lineno: err.lineNumber});
+					// continue at next game
+					stream.skipToNextGame();
+				}
 			}
-			games.push(currentPos);
 		}
-		return new Database({ text: pgnString, games: games, currentGameIndex: -1 }, gameCountGetterImpl, gameGetterImpl);
+		return new Database({ text: pgnString, games: games, currentGameIndex: -1, errors: errors }, gameCountGetterImpl, gameGetterImpl);
 	}
 
 	// Parse one game...
 	else {
 		var gameCounter = 0;
 		while(gameCounter < gameIndex) {
-			if(doSkipGame(stream)) {
+			if(checkValidGame(stream)) {
 				++gameCounter;
 			}
 			else {
-				throw new exception.InvalidPGN(pgnString, pgnString.length, i18n.INVALID_GAME_INDEX, gameIndex, gameCounter);
+				throw new exception.InvalidPGN(pgnString, pgnString.length, stream._lineCount, i18n.INVALID_GAME_INDEX, gameIndex, gameCounter);
 			}
 		}
 		return doParseGame(stream);
 	}
 };
+
+/**
+ * Write the date in PGN format
+ *
+ * @param {Date|{year:number, month:number}|{year:number}|undefined}} date
+ *
+ * @returns {string}
+ */
+function writeDate(date) {
+	if (date === undefined) {
+		return '[Date "????.??.??"]' + Separator;
+	} else if (date instanceof Date) {
+		return '[Date "' +
+			date.getFullYear().toString() + '.' +
+			((date.getMonth() + 1 < 10) ? ('0' + (date.getMonth() + 1).toString()) : (date.getMonth() + 1)).toString() + '.' +
+			((date.getDate() < 10) ? ('0' + date.getDate().toString()) : date.getDate().toString()) +
+			'"]' + Separator;
+	} else {
+		return '[Date "' +
+			((date.year === undefined) ? '????' : date.year) + '.' +
+			((date.month === undefined) ? '??' : date.month) + '.' +
+			((date.date === undefined) ? '??' : date.date) +
+			'"]' + Separator;
+	}
+}
+
+/**
+ * Write any additional headers not directly present as keys on game
+ * @param {Game} game
+ *
+ * @returns {string}
+ */
+function writeAdditionalHeaders(game) {
+	var res = '';
+	if (Object.keys(game.additionalHeaders()).length > 0) {
+		var headers = game.additionalHeaders();
+		for (var header in headers) {
+			if (Object.prototype.hasOwnProperty.call(headers, header)) {
+				if (!SupportedHeaders.includes(header)) {
+					res += ('[' + header + ' "' + headers[header] + '"]' + Separator);
+				}
+			}
+		}
+		/*Object.entries(game.additionalHeaders()).forEach((header, value) => {
+			if (!SupportedHeaders.includes(header)) {
+				res += ('[' + header + ' "' + value + '"]' + Separator);
+			}
+		});*/
+	}
+	return res;
+}
+
+/**
+ * Write the headers in order, first the standard ones, then any others
+ *
+ * @param {Game} game
+ *
+ * @returns {string}
+ */
+function writeHeaders(game) {
+	var res = '';
+	res += '[Event "' + ((game.event() === undefined) ? '?' : game.event()) + '"]' + Separator;
```

Comment:
The special characters `"` (double-quote) and `\` (backslash) must be escaped when serializing header values.

Reply:
Will do.